### ISSUE TITLE
DragDrop throwing COM error 'class not registered'

### ISSUE THREAD

**Issue by TheWerewolf** [AvaloniaUI/Avalonia#12497]
**Describe the bug**
When calling DragDrop.DoDragDrop() from inside a PointerMoved handler in debug mode, the call throws a COM exception.

**To Reproduce**
This is a control I'm developing. The idea is to drag-drop a tile from Actions to Tasks which are both ItemsControls.

![image](https://github.com/AvaloniaUI/Avalonia/assets/12217281/13945400-17bd-4993-99ac-49ad5593c7a1)

The source is the following XAML:

`<Border Grid.Row="1" Margin="0,5,0,0"  Background="GhostWhite">
<ScrollViewer VerticalScrollBarVisibility="Auto">
	<ItemsControl ItemsSource="{Binding Toolbox}" Background="GhostWhite" Margin="2">
		<ItemsControl.ItemTemplate>
			<DataTemplate>
				<Border CornerRadius="3" BorderBrush="LightSteelBlue" BorderThickness="1"
                                      Background="{Binding TileBackground}" Margin="0,0,0,3"
                                      **PointerMoved="Toolbox_MouseMove"**
                                      ToolTip.Tip="{Binding Description}"
                                      Tag="{Binding}">
					<Grid Height="32">
						<TextBlock Text="{Binding Name}" VerticalAlignment="Center"
							Margin="10,3,3,3" ToolTip.Tip="{Binding Description}"/>
					</Grid>
				</Border>
			</DataTemplate>
		</ItemsControl.ItemTemplate>
	</ItemsControl>
</ScrollViewer>
</Border>`

This is the code I'm using. 

       public async void Toolbox_MouseMove(object sender, PointerEventArgs e)
        {
            var pt = e.GetCurrentPoint(null);
            
            if (pt.Properties.IsLeftButtonPressed)
            {
                var ui = sender as Control;
                if (ui != null && ui.Tag != null && ui.Tag is IRuleItemCore ir)
                {
                    // Package the data.
                    DataObject data = new DataObject();
                    data.Set("IRuleItemCore", ir);
                    var fmts = data.GetDataFormats();

                    // Initiate the drag-and-drop operation.
                    await DragDrop.DoDragDrop(e, data, DragDropEffects.Copy | DragDropEffects.Move);
                }
            }
        }

As soon as the DoDragDrop is called I get:
![image](https://github.com/AvaloniaUI/Avalonia/assets/12217281/99b7f734-f8a6-4658-b68c-3147622bf62a)

Stack trace
   at Avalonia.Win32.Interop.UnmanagedMethods.DoDragDrop(IntPtr dataObject, IntPtr dropSource, Int32 allowedEffects, Int32& finalEffect)
   at Avalonia.Win32.DragSource.DoDragDrop(PointerEventArgs triggerEvent, IDataObject data, DragDropEffects allowedEffects)
   at Avalonia.Input.DragDrop.DoDragDrop(PointerEventArgs triggerEvent, IDataObject data, DragDropEffects allowedEffects)
   at DBControlsAV.ProgramUI.ProgramCore.RulePageCore.<Toolbox_MouseMove>d__59.MoveNext() in C:\Users\Werewolf\Documents\GitHub\DBControlsAV\DBControlsAV\ProgramUI\ProgramCore\RulePageCore.cs:line 421

I've also tried this using DataFormats.Text and a basic text string ("test') and it fails there too.

**Expected behavior**
The IRuleItemCore object should be passed to the drop target on the other column, but the drop method is never called.

**Screenshots**
See above.

**Desktop (please complete the following information):**

- OS: Windows 10
- Version: Avalonia UI 11.0.2

**Additional context**
Reran it in release mode and may have additional info:
![image](https://github.com/AvaloniaUI/Avalonia/assets/12217281/cbda8bb6-eef4-40ca-949a-e9ce01930a62)

**Comment by timunie** [comment]
check out Avalonia.Behaviors. There you can find working examples for it.

**Comment by TheWerewolf** [comment]
Reading through the source for Avalonia.Xaml.Behaviors/Avalonia.Xaml.Interactions.DragAndDrop and my use looks pretty much identical other than not bracketing the initial move distance. I added that in and it made no difference. 

I'd like to keep my code as close as possible to the original WPF version so gluing in classes and having to modify my XAML, VMs and code behind to fix what might be a bug in Avalonia seems counterproductive. (Then again, since the Drop event isn't an attached property for some reason... I already have to.)

And it really shouldn't throw a COM Exception - that should be caught by the Avalonia framework and rethrown with perhaps more helpful information. Trying to catch it won't make sense on other platforms that aren't using COM.

From the screenshot I included in the update, it looks more like an issue with the _wrapper object and looking over the change logs, by coincidence, there were mods to the _wrapper object in DragDrop on Monday this week, but I'm using the Nuget version, so... no easy way to check.

**Comment by TheWerewolf** [comment]
Found it.

The problem was the drop routine signature.

I had it as "public void Ruleset_Drop(DragEventArgs e)" when it should have been "public void Ruleset_Drop(object sender, DragEventArgs e)"

Because I started by trying to attach it as an event in Xaml and was getting weird error messages like 'cannot find setter with parameters like...' I misinterpreted the error to mean I had to remove the "object sender" (it seemed to be suggesting that there should only be a DragEventArgs parameter). 

I fixed all the other events which ARE attached (like PointerPressed) but because this one went through an AddHandler, which can't check the required type, it didn't catch the error and bang.

Ironically, the original WPF code was correct and I changed it in response to the error messages I was getting.

Anyway, closing the bug and just suggesting type checking and error messages can be improved.
